src/librustc_typeck/collect.rs

```diff
@@ -102,17 +102,115 @@ pub struct ItemCtxt<'tcx> {
 
 ///////////////////////////////////////////////////////////////////////////
 
+crate struct PlaceholderHirTyCollector(crate Vec<Span>);
+
+impl<'v> Visitor<'v> for PlaceholderHirTyCollector {
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'v> {
+        NestedVisitorMap::None
+    }
+    fn visit_ty(&mut self, t: &'v hir::Ty) {
+        if let hir::TyKind::Infer = t.kind {
+            self.0.push(t.span);
+        }
+        hir::intravisit::walk_ty(self, t)
+    }
+}
+
+impl PlaceholderHirTyCollector {
+    pub fn new() -> PlaceholderHirTyCollector {
+        PlaceholderHirTyCollector(vec![])
+    }
+}
+
 struct CollectItemTypesVisitor<'tcx> {
     tcx: TyCtxt<'tcx>,
 }
 
+crate fn placeholder_type_error(
+    tcx: TyCtxt<'tcx>,
+    ident_span: Span,
+    generics: &[hir::GenericParam],
+    placeholder_types: Vec<Span>,
+    suggest: bool,
+) {
+    if !placeholder_types.is_empty() {
+        let possible_names = ["T", "K", "L", "A", "B", "C"];
+        let used_names = generics.iter().filter_map(|p| match p.name {
+            hir::ParamName::Plain(ident) => Some(ident.name),
+            _ => None,
+        }).collect::<Vec<_>>();
+
+        let mut type_name = "ParamName";
+        for name in &possible_names {
+            if !used_names.contains(&Symbol::intern(name)) {
+                type_name = name;
+                break;
+            }
+        }
+        
+        let mut sugg: Vec<_> = placeholder_types.iter()
+            .map(|sp| (*sp, type_name.to_string()))
+            .collect();
+        if generics.is_empty() {
+            sugg.push((ident_span.shrink_to_hi(), format!("<{}>", type_name)));
+        } else {
+            sugg.push((
+                generics.iter().last().unwrap().span.shrink_to_hi(),
+                format!(", {}", type_name),
+            ));
+        }
+        let mut err = struct_span_err!(
+            tcx.sess,
+            placeholder_types.clone(),
+            E0121,
+            "the type placeholder `_` is not allowed within types on item signatures",
+        );
+        for span in &placeholder_types {
+            err.span_label(*span, "not allowed in type signatures");
+        }
+        if suggest {
+            err.multipart_suggestion(
+                "use type parameters instead",
+                sugg,
+                Applicability::HasPlaceholders,
+            );
+        }
+        err.emit();
+    }
+}
+
+fn reject_placeholder_type_signatures_in_item(tcx: TyCtxt<'tcx>, item: &'tcx hir::Item<'tcx>) {
+    let (generics, suggest) = match &item.kind {
+        hir::ItemKind::Union(_, generics)
+        | hir::ItemKind::Enum(_, generics)
+        | hir::ItemKind::Struct(_, generics) => (&generics.params[..], true),
+        hir::ItemKind::Static(ty, ..) => {
+            if let hir::TyKind::Infer = ty.kind {
```

Comment:
what about more complex infer types than just plain `_`?

Reply:
They result in a cycle error when trying to infer the types.